src/types/data/date.js

```diff
@@ -1,80 +1,168 @@
-  onChange: React.PropTypes.func.isRequired,
-  onBlur: React.PropTypes.func.isRequired
-};
+  componentWillReceiveProps(props) {
+    this.setState({input: createInitialState(props).input});
+  }
 
-function formatDateValue(value) {
-  if (typeof value == 'string') {
-    return moment.utc(value);
-  } else if (Number.isFinite(value)) {
-    return moment(value * 1000);
-  } else {
-    return '';
+  render() {
+    const {options, disabled} = this.props;
+    return (
+      <div
+        className='form-data-input-date-wrapper'
+        onFocus={this.handleFocus}
+        onBlur={this.handleBlur}
+      >
+        <input
+          className='form-data-input date'
+          type='text'
+          value={this.state.input}
+          onChange={this.handleInputChange}
+          placeholder={options.get('placeholder')}
+          disabled={disabled}
+        />
+        {this.state.showPicker && !disabled ? this.renderPicker() : null}
+      </div>
+    );
   }
-}
 
-function getDateFormat(dateType) {
-  switch (dateType) {
-    case 'time':
-      return 'hh:mm a';
-    case 'datetime':
-      return 'YYYY-MM-DD hh:mm a';
-    case 'date':
-    default:
-      return 'YYYY-MM-DD';
+  renderPicker() {
+    const {options, value} = this.props;
+    let datetime = unixToDatetime(value);
+    if (!datetime.isValid()) datetime = moment();
+    return (
+      <TetheredComponent
+        classes={{element: 'tether-element-top'}}
+        renderElementTo='body'
+        attachment='top left'
+      >
+        <div className='tether-target' />
+        <DatetimePicker
+          className='form-data-datetime-picker'
+          onFocusCapture={this.handleFocus}
+          onBlur={this.handleBlur}
+          moment={datetime}
+          onChange={this.handlePickerChange}
+          onDone={this.handleBlur}
+          type={options.get('dateType')}
+          prevMonthIcon='fa fa-angle-left'
+          nextMonthIcon='fa fa-angle-right'
+          doneIcon='fa fa-check'
+          dateIcon='fa fa-calendar'
+          timeIcon='fa fa-clock-o'
+        />
+      </TetheredComponent>
+    )
+  }
+
+  handleFocus = () => {
+    clearTimeout(this.blurTimeout);
+    this.setState({showPicker: true});
+  }
+
+  handleBlur = () => {
+    // Wait for possible focus event before handling the blur event.
+    this.blurTimeout = setTimeout(() => {
+      this.setState({showPicker: false}, () => {
+        this.saveInput();
+        this.props.onBlur();
+      });
+    });
+  }
+
+  saveInput = () => {
+    let unixTimestamp = null;
+    if (this.state.input) {
+      let datetime = stringToDatetime(this.state.input);
+      if (datetime.isValid()) {
+        unixTimestamp = datetimeToUnix(datetime);
+      }
+    }
+    if (unixTimestamp !== this.props.value) {
+      this.props.onChange(unixTimestamp);
+    }
+  }
+
+  handlePickerChange = (datetime) => {
+    const unixTimestamp = datetimeToUnix(datetime);
+    this.props.onChange(unixTimestamp);
```

Comment:
onChange will always get called because onBlur will always happen, unless the user types something in and then closes the page. I think this is pretty standard behavior for forms.

I've also changed onBlur to immediately save the input value (without waiting for the timeout), which should fix any edge cases where the user clicks submit after typing input.